src/main/java/com/monumental/services/MonumentService.java

```diff
@@ -19,34 +17,59 @@ public MonumentService(SessionFactoryService sessionFactoryService) {
+     * @param query         Your CriteriaQuery
+     * @param root          The root associated with your CriteriaQuery
+     * @param searchQuery   The string to search both fields for
+     */
+    private void buildFTSQuery(CriteriaBuilder builder, CriteriaQuery query, Root root, String searchQuery) {
+        query.where(
+            builder.or(
+                builder.equal(builder.function("fts", Boolean.class, root.get("title"), builder.literal(searchQuery)), true),
+                builder.equal(builder.function("fts", Boolean.class, root.get("artist"), builder.literal(searchQuery)), true)
+            )
+        );
+    }
+
+    /**
+     * Uses the FTS function as well as any filtering or pagination provided to search for matching Monuments
      */
     @SuppressWarnings("unchecked")
-    public List<Monument> search(String query) {
-        Session session = this.sessionFactoryService.getFactory().openSession();
-        Transaction transaction = null;
-        List<Monument> records;
-
-        try {
-            transaction = session.beginTransaction();
-            // The "fts" function is defined by the FTSFunction and CustomPostgreSQL9Dialect classes
-            Query q = session.createQuery(
-                    "select m from Monument m where fts(m.title, :query) = true"
-            ).setParameter("query", query);
-            records = q.list();
-            transaction.commit();
-            session.close();
-        } catch (HibernateException e) {
-            if (transaction != null) {
-                transaction.rollback();
-            }
-            session.close();
-            throw e;
+    public List<Monument> search(String searchQuery, String page, String limit) {
+
+        CriteriaBuilder builder = this.getCriteriaBuilder();
+        CriteriaQuery<Monument> query = this.createCriteriaQuery(builder, false);
+        Root<Monument> root = this.createRoot(query);
+        root.fetch("tags", JoinType.LEFT);
+        query.select(root).distinct(true);
+
+        // TODO: Query for tags
+        // TODO: Query for location
+        // TODO: Filters
+
+        if (searchQuery != null) {
+            this.buildFTSQuery(builder, query, root, searchQuery);
         }
 
-        return new ArrayList<>(new HashSet<>(records));
+        List<Monument> monuments = this.getWithCriteriaQuery(query, Integer.parseInt(limit), (Integer.parseInt(page)) - 1);
+
+        return monuments;
+    }
+
+    /**
+     * Count the total number of results for a FTS search
+     * TODO: Be sure to update this whenever the search function changes so that they stay in sync
+     */
+    public Integer countSearchResults(String searchQuery) {
+
+        CriteriaBuilder builder = this.getCriteriaBuilder();
+        CriteriaQuery<Long> query = builder.createQuery(Long.class);
+        Root<Monument> root = query.from(Monument.class);
+        query.select(builder.countDistinct(root)).distinct(true);
```

Comment:
Is doing `builder.countDistinct(root)).distinct(true)` necessary? Is this double the distinct-ness?

Reply:
I was losing my mind at the time trying to figure out why my count was always off. No, it's not necessary.

Turned out it was because in the search query I was using `.fetch("tags")` to load in the tags, but this was actually making it so only monuments with tags were returned while the count was returning those without tags too. The reason was because I had to set the join type of the fetch to `JoinType.LEFT` so that monuments without tags were returned